lib/linked_list.rb

```diff
@@ -12,150 +12,266 @@ def initialize(value, next_node = nil)
+  def get_at_index(index)
+    i = 0
+    current = @head
+    while current != nil
+      return current.data if index == i
+      i += 1
+      current = current.next
     end
+    return nil
+  end
 
-    # method to print all the values in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def visit
-      raise NotImplementedError
+  # method to print all the values in the linked list
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
+  def visit
+    current = @head
+    while current != nil
+      puts current.data
+      current = current.next
     end
+  end
 
-    # method to delete the first node found with specified value
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def delete(value)
-      raise NotImplementedError
+  # method to delete the first node found with specified value
+  # Time Complexity: O(1)
+  # Space Complexity: O(1)
+  def delete(value)
+    return nil if !@head
+    if @head.data == value
+      @head = @head.next
+      return
     end
-
-    # method to reverse the singly linked list
-    # note: the nodes should be moved and not just the values in the nodes
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def reverse
-      raise NotImplementedError
+    current = @head
+    while current != nil
+      if current.next.data == value
+        current.next = current.next.next
+        return
+      end
+      current = current.next
     end
+  end
 
+  # method to reverse the singly linked list
+  # note: the nodes should be moved and not just the values in the nodes
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
+  def reverse
+    prev = nil
+    next_node = nil
+    current = @head
+    while current != nil
+     
+      next_node = current.next
+      current.next = prev
+      prev = current
+      current = next_node
+    end
+    @head = prev
+  end
 
-    ## Advanced Exercises
-    # returns the value at the middle element in the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def find_middle_value
-      raise NotImplementedError
+  ## Advanced Exercises
+  # returns the value at the middle element in the singly linked list
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
+  def find_middle_value
+    return nil if !@head
+    slow = @head
+    fast = @head
+    while fast.next != nil
+      slow = slow.next
+      fast = fast.next.next
     end
+    return slow.data
+  end
 
-    # find the nth node from the end and return its value
-    # assume indexing starts at 0 while counting to n
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def find_nth_from_end(n)
-      raise NotImplementedError
+  # find the nth node from the end and return its value
+  # assume indexing starts at 0 while counting to n
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
+  def find_nth_from_end(n)
+    return nil if !@head
+    current = @head
+    index = 1
+    while current != nil
+      if length - index == n

```

Comment:
Just a note, that `length` is a method that traverses the list each time.  So this makes it O(n^2) time complexity. 

Better would be to save the length into a local variable and then use that local variable in the loop.